### ISSUE TITLE
Cause a transaction to roll back

### ISSUE THREAD

**Issue by remeike** [andremarianiello/hasql-transaction-io#3]
I've been trying to force a transaction to rollback given a certain condition. I noticed there isn't an equivalent function to `condemn` from `hasql-transaction`, so I tried to just issue a `ROLLBACK` from within `TransactionIO`. It actually worked just fine, but I've been getting a `there is no transaction in progress` warning, which is a little perplexing since it definitely did rollback a transaction. So I guess I was just wondering if there's a recommended way to do this with the library. Thanks for publishing this package!

**Comment by andremarianiello** [comment]
Hello! Thanks for trying out the library and filing my first issue! If I had to guess what is happening, I think that you are ending the transaction with your ROLLBACK, but then the library is ready to end the transaction it doesn't know that you already ended it, so it tries to end it again, giving that warning. The only way to for the library to know to rollback (right now) would to throw an exception from inside your TransactionIO. This triggers completion of the transaction via a rollback here https://github.com/andremarianiello/hasql-transaction-io/blob/e5a289f2a3283e620a571dc0b2c15d7b5983a54c/library/Hasql/Private/TransactionIO.hs#L74. Do you like this approach? If not I may be able to add a `condemn` function.

**Comment by remeike** [comment]
Oh, that makes a lot of sense. Throwing an exception works just fine for my purposes, so I can definitely take that approach!

A `condemn` function might be a nice addition, but I imagine that wouldn't be trivial to implement. I suppose you'd need something like a `StateT` to track whether the transaction was rolled back?

**Comment by andremarianiello** [comment]
Glad to hear it! I will keep this ticket open to remind me to look into `condemn` or in case throwing an exception doesn't work for you. I might just make `condemn` throw a specific exception which I then catch.

**Comment by worm2fed** [comment]
As I can see `ReleaseException` is deprecated and it's recommended to use `ReleaseExceptionWith SomeException`

**Comment by worm2fed** [comment]
it would be nice to have a function in package, instead of manually defining:
```
-- | Aborts 'TransactionIO'
condemn :: TransactionIO a
condemn = throwIO ReleaseException

instance Exception ReleaseType
```

**Comment by worm2fed** [comment]
@andremarianiello I can see you've created a PR. Many thanks!

while implementing my own version, I've noticed a strange behaviour.

```haskell
-- | Runs 'Sql.Transaction'. In case of error in any step of execution - it will rollback.
runTransaction :: With env m => Sql.Transaction e a -> m (Either e a)
runTransaction (Sql.runTransaction -> tx) =
  withFrozenCallStack
    ( runDb
        . transactionIO Serializable ReadWrite Deferrable
        $ runExceptT tx
          >>= either (\e -> Sql.condemn >> pure (Left e)) (pure . Right)
    )
{-# INLINE runTransaction #-}
```

It runs `ExceptT` and in case there was some domain level errors - it will rollback transaction and return an error.
Note, that `Sql.Transaction` defined like this:
```haskell
newtype Transaction error result = Transaction
  {runTransaction :: ExceptT error TransactionIO result}
  deriving newtype (Functor, Applicative, Monad, MonadIO, MonadError error)
````

Later - I'm checking that `Either` in return type, log that error and return appropriate http response.
So, when I'm using `condemn` - my logs is like this.
```
[Info]    [18 May 2023 10:44:36.909 +00:00] [Context.Domain.Security.Logic.RequestCode.requestCode#59] [ThreadId 17] ... requested Otp to sign in
ReleaseException
```

and without `condemn`
```
[Info]    [18 May 2023 10:52:07.283 +00:00] [Context.Domain.Security.Logic.RequestCode.requestCode#59] [ThreadId 17] ... requested Otp to sign in
[Info]    [18 May 2023 10:52:07.291 +00:00] [Context.Domain.Security.Logic.RequestCode.requestCode#66] [ThreadId 17] ToManyRequests 177
POST /signIn/request
  Request Body: {
  "email": "..."
}
  Accept: application/json;charset=utf-8
  Status: 429 Too Many Requests 0.009413s
```

I understand that here some specific implementation involved, but I suspect that `TransactionIO` do not catch exception properly, or I might be doing something wrong...

UPD: 
So, with your update I will get `CondemnTransactionException` instead of `ReleaseException`.
My expectation was, like this. Okay, exception is used to rollback transaction, but when I rollback - I do not want that exception thrown in outer level. Like a command

**Comment by worm2fed** [comment]
I could solve it like this:
```haskell
runTransaction :: With env m => Sql.Transaction e a -> m (Either e a)
runTransaction (Sql.runTransaction -> tx) =
  withFrozenCallStack
    ( runDb
        . transactionIO Serializable ReadWrite Deferrable
        $ runExceptT tx
          >>= either (\e -> Sql.condemn `catch` \(_ :: SomeException) -> pure $ Left e) (pure . Right)
    )
{-# INLINE runTransaction #-}
```

But with this solution it do not rollback(

**Comment by worm2fed** [comment]
the best I can do here is define my internal errors as exceptions, in `runTransaction` just throw internal error as exception (from `Either`) and at another layer - catch that exception and turn back to `Either`...
Like this:
```haskell
runTransaction
  :: (With env m, Exception e, MonadUnliftIO m)
  => Sql.Transaction e a
  -> m (Either e a)
runTransaction (Sql.runTransaction -> tx) = withFrozenCallStack do
  let action =
        runDb
          . transactionIO Serializable ReadWrite Deferrable
          $ runExceptT tx >>= fromEither
  (Right <$> action) `catch` (pure . Left)
{-# INLINE runTransaction #-}
```

**Comment by andremarianiello** [comment]
Thanks for the feedback @worm2fed ! According to [this comment](https://github.com/andremarianiello/hasql-transaction-io/issues/3#issuecomment-1549760317) you wanted a `condemn` function that simply throws an exception, but it sounds like that behavior is not exactly what you want. Could you explain exactly how you want `condemn` to behave?

**Comment by worm2fed** [comment]
@andremarianiello sorry for confusion.
First I've implemented condemn like you suggested, simply throwing specific exception.

But this implementation has the described above behavior.

With my example, I use `TransactionIO` wrapped to `ExceptT`, which allows me to create a chains of transactions with possibility to fail with domain specific errors. So when I run computation, I get `Either`. And in case of `Left` I want to be sure that transaction will be rollbacked, but I still need that `Either` to later translate domain specific errors to global application errors, then log and send appropriate http response.
So I would like to use condemn just as a command, but still have my response. 
It can lead to question, what if `TransactionIO` is not wrapped, so what should be return type when we can get it only after execution.. I understand it can be tricky.

I solved this problem by throwing exception and then catching it to turn back to `Either`. And I'm fine with it.
Just decided to explain a problem I faced, so it can either help someone or lead you to some useful thoughts.
